### DISCUSSION TITLE
Loading of static assets for disabled extensions

### DISCUSSION THREAD

**Discussion by Raghav-254** [jupyterlab/jupyterlab#16082] [Q&A]
Hello team,
As per our earlier discussion in this thread: https://github.com/jupyterlab/jupyterlab/discussions/15945, I was trying to validate whether the static assets are getting loaded for disabled extensions in jupyterlab 4. So for that I performed the following steps:
1. Upgrade to jupyterlab 4 via **pip install jupyterlab**.
2. Install a pip based open source extension [jupyter-ai](https://github.com/jupyterlab/jupyter-ai) via **pip install jupyter-ai**
3. Reloaded the page, and saw that **12** static JS files are getting loaded each time on page reload.
4. Now, I disabled this lab extension via **jupyter labextension disable @jupyter-ai/core**.
5. Now I reloaded the page again, and this time saw **11** static JS files are getting loaded each time on page reload. Ideally it should be **0** right as I have disabled the extension.

Can you please help me in understanding this behaviour? If disabling the extensions is not reducing the number of static JS files, that is getting loaded on page-reload, then what benefit this disabling of extensions functionality is offering us?

Best,
Raghav Mittal

**Comment by krassowski** [comment]
> Now, I disabled this lab extension via `jupyter labextension disable @jupyter-ai/core`.

Or did you? I would not recommend using the CLI for this because it is error-prone. In this instance jupyter-ai does not follow the established practice of prefixing plugin names with the extension name, so I think you may have not achieved the result you wanted. Instead, jupyter-ai (as of version 2.12) has three plugins named `jupyter_ai:plugin`, `jupyter_ai:inline-completions` and `jupyter_ai:status-item`. I would suspect your command may have had no effect (yes, `jupyter labextension list` output is misleading here).

Instead I would suggest trying out the Plugin Manager: https://jupyterlab.readthedocs.io/en/latest/user/extensions.html#managing-plugins-with-plugin-manager

> Reloaded the page, and saw that 12 static JS files are getting loaded each time on page reload.

This number seems off. There are way more JS files loaded on page reload. Which 12/11 files do you refer to? How do you defined "static" files in this context? Most of the extensions are loaded dynamically so it would not make sense to look at static files really.

**Reply by Raghav-254** [reply]
Hi @krassowski, you are right that `jupyter labextension disable @jupyter-ai/core`, has no effect as I was able to see the extension on the left side panel. I was unable to locate the plugin manager UI mentioned in the doc, so I ran the following three commands to disable the individual plugins:
1. jupyter labextension disable jupyter_ai:plugin
2. jupyter labextension disable jupyter_ai:inline-completions
3. jupyter labextension disable jupyter_ai:status-item
as it is written that in the doc that **Functionally, toggling an extension from the plugin manager is equivalent to running the jupyter labextension enable or jupyter labextension disable commands**.

After running all these commands and on refreshing the page, I didn't see the extension on the left side panel which indicates that the extension got disabled cleanly but I was still seeing the 11 static JS files corresponding to this jupyter_ai labextension that is getting loaded each time on page reload. Ideally it should not be loaded as I have disabled the extension now. Can you please confirm if this is the expected behaviour? If yes, then what benefit does this disabling of extension giving us?

Also I am using jupyterlab:v4.1.5, I forgot to mention this earlier.

Attaching the video as well which displays the 11 static JS files corresponding to this **jupyter_ai** labextension that are getting loaded on page-reload after disabling the individual plugins of it. I am calling the mentioned JS files in the video as **static** as they are not changing and getting picked from the disk cache.

https://github.com/jupyterlab/jupyterlab/assets/52010424/46c64e62-cc19-4ba4-a8ff-453e60224eba

**Comment by Raghav-254** [comment]
Hi @krassowski, could you kindly respond?

**Comment by Raghav-254** [comment]
Hi @krassowski, could you please kindly respond?

**Comment by krassowski** [comment]
> Attaching the video as well which displays the 11 static JS files corresponding to this `jupyter_ai` labextension that are getting loaded on page-reload after disabling the individual plugins of it. I am calling the mentioned JS files in the video as static as they are not changing and getting picked from the disk cache.

These files appear are the code of the extension. By _static asset_ I would understand things like CSS, which is not loaded for disabled extensions since https://github.com/jupyterlab/jupyterlab/pull/11962. As for not loading the source when disabling individual plugins - this may be possible but may require some changes in webpack bundling and other parts of the infrastructure.

> Can you please confirm if this is the expected behaviour?

Well, it depends on what you mean by "expected". Expected by the users? Expected as "this is what was the intent of the author of the code"? For the former I guess it is not what you expect. For the latter I can only guess.

> If yes, then what benefit does this disabling of extension giving us?

I would say of of today a couple:
- it does not fetch CSS assets (if coded properly)
- it does not execute the logic for disabled plugins which makes the startup faster and uses up less memory

Is it ideal? Probably not. The question is how can you/@linkedin contribute to make it better.

I would suggest to start by opening an issue with a feature request to prevent the code of disabled plugins from loading. I can then respond with more thoughts on what technical challenges would need to be addressed.

**Comment by Raghav-254** [comment]
Thanks for your reply @krassowski. Based on your above suggestion, have raised this feature request: https://github.com/jupyterlab/jupyterlab/issues/16135
